core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsMetricsManagerImpl.java

```diff
@@ -0,0 +1,400 @@
+
+   @Override
+   public void recordMessageSent(Address destination, int bytesSent, boolean async) {
+      if (stopped) {
+         return;
+      }
+      var metrics = get(destination);
+      if (metrics == null) {
+         return;
+      }
+      metrics.incrementBytesSent(bytesSent);
+      if (async) {
+         metrics.incrementAsyncRequests();
+      }
+   }
+
+   @Override
+   public synchronized void onChannelConnected(JChannel channel, boolean isMainChannel) {
+      if (stopped) {
+         return;
+      }
+      String nodeName = Objects.requireNonNull(nodeName(channel));
+      String clusterName = Objects.requireNonNull(channel.clusterName());
+      if (isMainChannel) {
+         assert mainChannelRegistry == null;
+         mainChannelRegistry = new MainChannelRegistry(nodeName, clusterName);
+      }
+      if (otherChannels.stream().map(m -> m.channel).noneMatch(ch -> ch.equals(channel))) {
+         otherChannels.add(new ClusterMetrics(channel));
+      }
+      if (mainChannelRegistry != null) {
+         otherChannels.forEach(clusterMetrics -> clusterMetrics.register(mainChannelRegistry));
+      }
+   }
+
+   @Override
+   public synchronized void onChannelDisconnected(JChannel channel) {
+      if (stopped) {
+         return;
+      }
+      if (mainChannelRegistry != null && mainChannelRegistry.clusterName.equals(channel.clusterName()) && mainChannelRegistry.nodeName.equals(channel.address().toString())) {
+         mainChannelRegistry = null;
+      }
+      Optional<ClusterMetrics> optMetrics = otherChannels.stream().filter(m -> m.channel.equals(channel)).findFirst();
+      if (optMetrics.isEmpty()) {
+         return;
+      }
+      ClusterMetrics metrics = optMetrics.get();
+      metrics.unregister(registry);
+      otherChannels.remove(metrics);
+   }
+
+   private DestinationMetrics get(Address dst) {
+      assert dst != null;
+      return perDestinationMetrics.computeIfAbsent(dst, this::createDestinationMetrics);
+   }
+
+   private DestinationMetrics createDestinationMetrics(Address destination) {
+      assert destination != null;
+      var statsRegistry = mainChannelRegistry;
+      if (statsRegistry == null) {
+         return null;
+      }
+
+      var attributes = createAttributes(destination.toString(), histogramEnabled);
+
+      // DestinationMetricsBuilder stores the references for timers/counters/etc.
+      DestinationMetricsBuilder builder = new DestinationMetricsBuilder();
+      // registerMetrics sets all the fields
+      var metricsIds = statsRegistry.registerStats(builder, attributes);
+      // create DestinationMetrics
+      return builder.build(metricsIds, histogramEnabled);
+   }
+
+   private static String nodeName(JChannel channel) {
+      org.jgroups.Address addr = channel.address();
+      return addr == null ? channel.name() : addr.toString();
+   }
+
+   private static Collection<MBeanMetadata.AttributeMetadata> createAttributes(String dst, boolean histogramEnabled) {
```

Comment:
No need to pass the `histogramEnabled` param as the class variable is always passed.